django_mongodb_backend/expressions/search.py

```diff
@@ -0,0 +1,1029 @@
+from django.db import NotSupportedError
+from django.db.models import CharField, Expression, FloatField, TextField
+from django.db.models.expressions import F, Value
+from django.db.models.lookups import Lookup
+
+from ..query_utils import process_lhs, process_rhs
+
+
+def cast_as_value(value):
+    if value is None:
+        return None
+    return Value(value) if not hasattr(value, "resolve_expression") else value
+
+
+def cast_as_field(path):
+    return F(path) if isinstance(path, str) else path
+
+
+class Operator:
+    AND = "AND"
+    OR = "OR"
+    NOT = "NOT"
+
+    def __init__(self, operator):
+        self.operator = operator
+
+    def __eq__(self, other):
+        if isinstance(other, str):
+            return self.operator == other
+        return self.operator == other.operator
+
+    def negate(self):
+        if self.operator == self.AND:
+            return Operator(self.OR)
+        if self.operator == self.OR:
+            return Operator(self.AND)
+        return Operator(self.operator)
+
+    def __hash__(self):
+        return hash(self.operator)
+
+    def __str__(self):
+        return self.operator
+
+    def __repr__(self):
+        return self.operator
+
+
+class SearchCombinable:
+    def _combine(self, other, connector):
+        if not isinstance(self, CompoundExpression | CombinedSearchExpression):
+            lhs = CompoundExpression(must=[self])
+        else:
+            lhs = self
+        if other and not isinstance(other, CompoundExpression | CombinedSearchExpression):
+            rhs = CompoundExpression(must=[other])
+        else:
+            rhs = other
+        return CombinedSearchExpression(lhs, connector, rhs)
+
+    def __invert__(self):
+        return self._combine(None, Operator(Operator.NOT))
+
+    def __and__(self, other):
+        return self._combine(other, Operator(Operator.AND))
+
+    def __rand__(self, other):
+        return self._combine(other, Operator(Operator.AND))
+
+    def __or__(self, other):
+        return self._combine(other, Operator(Operator.OR))
+
+    def __ror__(self, other):
+        return self._combine(other, Operator(Operator.OR))
+
+
+class SearchExpression(SearchCombinable, Expression):
+    """
+    Base expression for MongoDB Atlas `$search` stage.
+
+    Subclasses produce the operator document placed under $search and expose
+    the stage to queryset methods such as annotate(), filter(), or order_by().
+    """
+
+    output_field = FloatField()
+
+    def __str__(self):
+        cls = self.identity[0]
+        kwargs = dict(self.identity[1:])
+        arg_str = ", ".join(f"{k}={v!r}" for k, v in kwargs.items())
+        return f"<{cls.__name__}({arg_str})>"
```

Comment:
Is it tested? (I'm trying to get a coverage report working but I didn't spot any `str(` in tests.)